pkg/provider/cloud/kubevirt/csi.go

```diff
@@ -0,0 +1,163 @@
+		return name, func(sa *corev1.ServiceAccount) (*corev1.ServiceAccount, error) {
+			return sa, nil
+		}
+	}
+}
+
+func csiSecretTokenCreator(name string) reconciling.NamedSecretCreatorGetter {
+	return func() (string, reconciling.SecretCreator) {
+		return name, func(s *corev1.Secret) (*corev1.Secret, error) {
+			s.Annotations[corev1.ServiceAccountNameKey] = csiResourceName
+			s.Type = corev1.SecretTypeServiceAccountToken
+			return s, nil
+		}
+	}
+}
+
+func csiRoleCreator(name string) reconciling.NamedRoleCreatorGetter {
+	return func() (string, reconciling.RoleCreator) {
+		return name, func(r *rbacv1.Role) (*rbacv1.Role, error) {
+			r.Rules = []rbacv1.PolicyRule{
+				{
+					APIGroups: []string{"cdi.kubevirt.io"},
+					Resources: []string{"datavolumes"},
+					Verbs:     []string{"get", "create", "delete"},
+				},
+				{
+					APIGroups: []string{"kubevirt.io"},
+					Resources: []string{"virtualmachineinstances"},
+					Verbs:     []string{"list"},
+				},
+				{
+					APIGroups: []string{"subresources.kubevirt.io"},
+					Resources: []string{"virtualmachineinstances/addvolume", "virtualmachineinstances/removevolume"},
+					Verbs:     []string{"update"},
+				},
+			}
+
+			return r, nil
+		}
+	}
+}
+
+func csiRoleBindingCreator(name, namespace string) reconciling.NamedRoleBindingCreatorGetter {
+	return func() (string, reconciling.RoleBindingCreator) {
+		return name, func(rb *rbacv1.RoleBinding) (*rbacv1.RoleBinding, error) {
+			rb.Subjects = []rbacv1.Subject{
+				{
+					Kind:      "ServiceAccount",
+					Name:      name,
+					Namespace: namespace,
+				},
+			}
+
+			rb.RoleRef = rbacv1.RoleRef{
+				APIGroup: "rbac.authorization.k8s.io",
+				Kind:     "Role",
+				Name:     name,
+			}
+
+			return rb, nil
+		}
+	}
+}
+
+// reconcileCSIRoleRoleBinding reconciles the Role and RoleBinding needed by CSI driver.
+func reconcileCSIRoleRoleBinding(ctx context.Context, namespace string, client ctrlruntimeclient.Client, restConfig *restclient.Config) error {
+	roleCreators := []reconciling.NamedRoleCreatorGetter{
+		csiRoleCreator(csiResourceName),
+	}
+	if err := reconciling.ReconcileRoles(ctx, roleCreators, namespace, client); err != nil {
+		return err
+	}
+
+	roleBindingCreators := []reconciling.NamedRoleBindingCreatorGetter{
+		csiRoleBindingCreator(csiResourceName, csiServiceAccountNamespace),
+	}
+	if err := reconciling.ReconcileRoleBindings(ctx, roleBindingCreators, namespace, client); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func EnsureCSIInfraTokenAccess(ctx context.Context, infraKubeconfig string) ([]byte, error) {
+	client, restConfig, err := NewClientWithRestConfig(infraKubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	saCreators := []reconciling.NamedServiceAccountCreatorGetter{
+		csiServiceAccountCreator(csiResourceName),
+	}
+	if err := reconciling.ReconcileServiceAccounts(ctx, saCreators, csiServiceAccountNamespace, client); err != nil {
```

Comment:
Do we take the opportunity of this PR to create the sa in the correct namespace ?

Reply:
I would prefer to not make this change in this PR. It will be a little bit bigger change.

Reply:
Fine for me.